src/Symfony/Bundle/DebugBundle/VarDumper/DelayedDumper.php

```diff
@@ -0,0 +1,121 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\DebugBundle\VarDumper;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use Symfony\Component\VarDumper\Cloner\Data;
+use Symfony\Component\VarDumper\Dumper\CliDumper;
+use Symfony\Component\VarDumper\Dumper\DataDumperInterface;
+use Symfony\Component\VarDumper\Dumper\HtmlDumper;
+
+/**
+ * Used as fallback in the ServerDumper, it delays output to the end of the process for web requests.
+ * It replicates the DumpDataCollector behavior to ensure dumps are written to the output whenever the dump server isn't up.
+ * In cli, output is not delayed and the provided CliDumper instance is used instead.
+ *
+ * @author Maxime Steinhausser <maxime.steinhausser@gmail.com>
+ *
+ * @internal
+ */
+final class DelayedDumper implements DataDumperInterface, EventSubscriberInterface
+{
+    private $charset;
+    private $flags;
+    private $displayOptions;
+
+    /** @var Data[] */
+    private $data = array();
+
+    /** @var CliDumper */
+    private $cliDumper;
+
+    public function __construct(CliDumper $cliDumper, string $charset = null, int $flags = 0, array $displayOptions = array())
+    {
+        $this->charset = $charset;
+        $this->flags = $flags;
+        $this->displayOptions = $displayOptions;
+        $this->cliDumper = $cliDumper;
+    }
+
+    public function dump(Data $data)
+    {
+        // only delay web requests
+        if (!\in_array(PHP_SAPI, array('cli', 'phpdbg'), true)) {
```

Comment:
I would use `if (\PHP_SAPI === 'cli' || \PHP_SAPI === 'phpdbg')`, which can be resolved by OPCache at compile time